boot/bootutil/include/bootutil/fault_injection_hardening.h

```diff
@@ -0,0 +1,339 @@
+ * as otherwise the CFI counter will not be decremented and the CFI check will
+ * fail causing a panic.
+ */
+
+#include "mcuboot_config/mcuboot_config.h"
+
+#if defined(MCUBOOT_FIH_PROFILE_HIGH)
+
+#define FIH_ENABLE_DELAY         /* Requires an entropy source */
+#define FIH_ENABLE_DOUBLE_VARS
+#define FIH_ENABLE_GLOBAL_FAIL
+#define FIH_ENABLE_CFI
+
+#elif defined(MCUBOOT_FIH_PROFILE_MEDIUM)
+
+#define FIH_ENABLE_DOUBLE_VARS
+#define FIH_ENABLE_GLOBAL_FAIL
+#define FIH_ENABLE_CFI
+
+#elif defined(MCUBOOT_FIH_PROFILE_LOW)
+
+#define FIH_ENABLE_GLOBAL_FAIL
+#define FIH_ENABLE_CFI
+
+#else
+#define MCUBOOT_FIH_PROFILE_OFF
+#endif /* MCUBOOT_FIH_PROFILE */
+
+#ifdef FIH_ENABLE_DELAY
+#include "fault_injection_hardening_delay_rng.h"
+#endif /* FIH_ENABLE_DELAY */
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+/* Non-zero success value to defend against register resets. Zero is the most
+ * common value for a corrupted register so complex bit-patterns are used
+ */
+#ifndef MCUBOOT_FIH_PROFILE_OFF
+#define FIH_POSITIVE_VALUE 0x1AAAAAAA
+#define FIH_NEGATIVE_VALUE 0x15555555
+#else
+#define FIH_POSITIVE_VALUE 0
+#define FIH_NEGATIVE_VALUE 1
+#endif
+
+/* A volatile mask is used to prevent compiler optimization - the mask is xored
+ * with the variable to create the backup and the integrity can be checked with
+ * another xor. The mask value doesn't _really_ matter that much, as long as
+ * it has reasonably high hamming weight.
+ */
+#define _FIH_MASK_VALUE 0xBEEF
+
+#ifdef FIH_ENABLE_DOUBLE_VARS
+
+/* All ints are replaced with two int - the normal one and a backup which is
+ * XORed with the mask.
+ */
+extern volatile int _fih_mask;
+typedef volatile struct {
+    volatile int val;
+    volatile int msk;
+} fih_int;
+
+#else
+
+typedef int fih_int;
+
+#endif /* FIH_ENABLE_DOUBLE_VARS */
+
+extern fih_int FIH_SUCCESS;
+extern fih_int FIH_FAILURE;
+
+#ifdef FIH_ENABLE_GLOBAL_FAIL
+/* Global failure handler - more resistant to unlooping. noinline and used are
+ * used to prevent optimization
+ */
+__attribute__((noinline)) __attribute__((used))
+void fih_panic_loop(void);
+#define FIH_PANIC fih_panic_loop()
+#else
+#define FIH_PANIC while (1) {}
+#endif  /* FIH_ENABLE_GLOBAL_FAIL */
+
+/* NOTE: For functions to be inlined outside their compilation unit they have to
+ * have the body in the header file. This is required as function calls are easy
+ * to skip.
+ */
+#ifdef FIH_ENABLE_DELAY
+#include "bootutil/bootutil_log.h"
+/* Delaying logic, with randomness from a CSPRNG */
+__attribute__((always_inline)) inline
+int fih_delay(void)
+{
+    unsigned char delay;
+    int foo = 0;
+    volatile int rc;
+
+    delay = fih_delay_random_uchar();
+
+    for (volatile int i = 0; i < delay; i++) {
+        foo++;
+    }
+
+    rc = 1;
+
+    /* rc is volatile so if it is the return value then the function cannot be
+     * optimized
+     */
+    return rc;
+}
+#else
+__attribute__((always_inline)) inline
```

Comment:
minor: The `always inline` attribute is still an indication that a function must be inlined. Macros are the better way to ensure inlining. Same comment below.

Reply:
There's a deliberate tradeoff between usability and performance going on here; Macros are certainly be best way to ensure things are always inlined but they come with some tradeoffs. They tend to cause issues with debuggers / debug symbols and also things like code-completion / code formatters etc. All the compilers I tested did respect the `always_inline`, but I do agree that it could be an issue in theory.

Reply:
Your are right. So far they were inlined. I would keep in this way for simplicity. If one want to use this feature then the compiled code must be verified by test and visual inspection. So functions not be inlined should be caught by these steps.